### PR TITLE
Fix #3225 proactive-client pause() scheduler does not pause the sched…

### LINKED ISSUE #3225: proactive-client pause() scheduler does not pause the scheduler
Either via non interactive (./bin/proactive-client -pause) or via interactive (pause()) mode , the scheduler is not paused.

After pausing the scheduler the jobs are still submitted and executed

### PR DESCRIPTION
**Pull request by mbenguig**
…uler

### CODE REVIEW

File: rest/rest-server/src/test/java/functionaltests/RestfulSchedulerPauseTest.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    public void testPauseScheduler() throws Exception {
+        String resourceUrl = getResourceUrl("pause");
+        HttpPut httpPut = new HttpPut(resourceUrl);
+        setSessionHeader(httpPut);
+        HttpResponse response = executeUriRequest(httpPut);
+        assertHttpStatusOK(response);
+        assertTrue(Boolean.valueOf(getContent(response)));
+        Scheduler scheduler = RestFuncTHelper.getScheduler();
+        assertEquals(SchedulerStatus.PAUSED, scheduler.getStatus());
+    }
```
Comment (Mykhailenko) [inline rest/rest-server/src/test/java/functionaltests/RestfulSchedulerPauseTest.java]: Is there a test which checks that pause mechanism actually works?

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/LiveJobs.java [hunk 2]
```diff
@@ -298,7 +298,7 @@ void jobSubmitted(InternalJob job) {
         listener.jobSubmitted(clientJobState);
     }
 
-    Map<JobId, JobDescriptor> lockJobsToSchedule() {
+    Map<JobId, JobDescriptor> lockJobsToSchedule(boolean isSchedulerPaused) {
```
Comment (Mykhailenko) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/LiveJobs.java]: I know that is not API method, however, could you please add javadoc which explains what happen when isSchedulerPaused true or false. thanks.

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java [hunk 3]
```diff
@@ -187,6 +187,10 @@ public boolean isSubmitPossible() {
         return status.isSubmittable();
     }
 
+    public boolean isPaused() {
+        return status == SchedulerStatus.PAUSED;
```
Comment (fviale) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java]: should apply also to SchedulerStatus.STOPPED (change the method name accordingly)

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/LiveJobs.java [hunk 4]
```diff
@@ -307,18 +307,20 @@ void jobSubmitted(InternalJob job) {
         for (Map.Entry<JobId, JobData> entry : jobs.entrySet()) {
             JobData value = entry.getValue();
 
+            // If the scheduler is paused, schedule only running jobs
+            if (isSchedulerPaused && value.job.getStatus() != JobStatus.RUNNING) {
```
Comment (fviale) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/LiveJobs.java]: should apply also to JobStatus.STALLED

File: rest/rest-server/src/test/java/functionaltests/RestfulSchedulerPauseTest.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+import org.apache.http.HttpResponse;
+import org.apache.http.client.methods.HttpPut;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.ow2.proactive.scheduler.common.Scheduler;
+import org.ow2.proactive.scheduler.common.SchedulerStatus;
+
+
+public class RestfulSchedulerPauseTest extends AbstractRestFuncTestCase {
```
Comment (fviale) [inline rest/rest-server/src/test/java/functionaltests/RestfulSchedulerPauseTest.java]: A functional test in scheduler/scheduler-server should also be created to verify the behavior (when the scheduler is paused, only running or stalled jobs will terminate.

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 9 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [LiveJobs.java#L698](https://github.com/ow2-proactive/scheduling/blob/6b3d2bc4e6ae0f0eb4baace2d431b770a3bdadb0/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/LiveJobs.java#L698): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [LiveJobs.java#L781](https://github.com/ow2-proactive/scheduling/blob/6b3d2bc4e6ae0f0eb4baace2d431b770a3bdadb0/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/LiveJobs.java#L781): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [LiveJobs.java#L836](https://github.com/ow2-proactive/scheduling/blob/6b3d2bc4e6ae0f0eb4baace2d431b770a3bdadb0/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/LiveJobs.java#L836): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [LiveJobs.java#L872](https://github.com/ow2-proactive/scheduling/blob/6b3d2bc4e6ae0f0eb4baace2d431b770a3bdadb0/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/LiveJobs.java#L872): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L607](https://github.com/ow2-proactive/scheduling/blob/6b3d2bc4e6ae0f0eb4baace2d431b770a3bdadb0/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L607): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L634](https://github.com/ow2-proactive/scheduling/blob/6b3d2bc4e6ae0f0eb4baace2d431b770a3bdadb0/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L634): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L661](https://github.com/ow2-proactive/scheduling/blob/6b3d2bc4e6ae0f0eb4baace2d431b770a3bdadb0/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L661): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L688](https://github.com/ow2-proactive/scheduling/blob/6b3d2bc4e6ae0f0eb4baace2d431b770a3bdadb0/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L688): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L713](https://github.com/ow2-proactive/scheduling/blob/6b3d2bc4e6ae0f0eb4baace2d431b770a3bdadb0/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L713): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by mbenguig** [comment]
Sorry, there are some files that are not supposed to be in this PR. I have to remove this PR, and make e new one.

File: scheduler/scheduler-server/src/test/java/functionaltests/job/TestJobWhenSchedulerPaused.java [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+        // Ensure that every job newly submitted will be pending
+        JobId jobId1 = schedulerHelper.submitJob(new File(simpleJobUrl.toURI()).getAbsolutePath());
+        JobId jobId2 = schedulerHelper.submitJob(new File(simpleJobUrl.toURI()).getAbsolutePath());
+
+        schedulerHelper.waitForEventJobSubmitted(jobId1, EVENT_TIMEOUT);
+        schedulerHelper.waitForEventJobSubmitted(jobId2, EVENT_TIMEOUT);
+
+        // Without sleeping, we could could test if jobs are blocked in pending state ..
+        // .. without waiting long enough that they run
+        Thread.sleep(5000);
```
Comment (sonarqube-activeeon) [inline scheduler/scheduler-server/src/test/java/functionaltests/job/TestJobWhenSchedulerPaused.java]: MAJOR Remove this use of "Thread.sleep()". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2925)

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 10 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 9 minor

Watch the comments in this conversation to review them.

#### 9 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [LiveJobs.java#L705](https://github.com/ow2-proactive/scheduling/blob/af630fa11358df59a85879b07ee6de50063408e5/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/LiveJobs.java#L705): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [LiveJobs.java#L788](https://github.com/ow2-proactive/scheduling/blob/af630fa11358df59a85879b07ee6de50063408e5/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/LiveJobs.java#L788): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [LiveJobs.java#L843](https://github.com/ow2-proactive/scheduling/blob/af630fa11358df59a85879b07ee6de50063408e5/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/LiveJobs.java#L843): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [LiveJobs.java#L879](https://github.com/ow2-proactive/scheduling/blob/af630fa11358df59a85879b07ee6de50063408e5/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/LiveJobs.java#L879): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L607](https://github.com/ow2-proactive/scheduling/blob/af630fa11358df59a85879b07ee6de50063408e5/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L607): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L634](https://github.com/ow2-proactive/scheduling/blob/af630fa11358df59a85879b07ee6de50063408e5/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L634): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L661](https://github.com/ow2-proactive/scheduling/blob/af630fa11358df59a85879b07ee6de50063408e5/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L661): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L688](https://github.com/ow2-proactive/scheduling/blob/af630fa11358df59a85879b07ee6de50063408e5/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L688): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L713](https://github.com/ow2-proactive/scheduling/blob/af630fa11358df59a85879b07ee6de50063408e5/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L713): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request closed without merging**